1. Add tier-1 page allocator

niklas-arm opened the pull request:
This adds the page allocator algorithm, initialization code and SVC calls.

This is part of larger changes, but I'd prefer to review these parts separately when possible.

@Patater @AlessandroA @meriac

niklas-arm commented:
Note, that the page heap has to be added to the target linkerscript for the uvisor-header to link correctly.

meriac commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,239 @@
+#   define g_active_box 0
+#   define vmpu_is_box_id_valid(...) 0
+#   define HALT_ERROR(id, ...) DPRINTF(__VA_ARGS__)
+static uint32_t vmpu_unpriv_uint32_read(uint32_t addr)
```
@niklas-arm please add comment on replacing these two functions with the real unpriv functions headerfile from uVisor.

meriac commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,239 @@
+void page_allocator_init(void * const heap_start, void * const heap_end, uint32_t page_size)
+{
+    /* Make sure the UVISOR_PAGE_UNUSED is definitely NOT a valid box id! */
+    if (vmpu_is_box_id_valid(UVISOR_PAGE_UNUSED)) {
```
@niklas-arm coolness.

niklas-arm commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,240 @@
+#   define vmpu_is_box_id_valid(...) 0
+#   define HALT_ERROR(id, ...) DPRINTF(__VA_ARGS__)
+/* FIXME: Use real functions from `vmpu_unpriv_access.h` for unsupported too! */
+static uint32_t vmpu_unpriv_uint32_read(uint32_t addr)
```
@AlessandroA This would require making `vmpu_unpriv_access.h` available in the API header files.

AlessandroA replied:
Done. See #228

AlessandroA commented on core/system/inc/page_allocator.h:
```diff
@@ -0,0 +1,51 @@
+
+/* Initialize the page allocator.
+ *
+ * The most importent argument to this function is the page size. It determines
```
s/importent/important

AlessandroA commented on api/inc/halt_exports.h:
```diff
@@ -17,9 +17,15 @@
+#define UVISOR_ERROR_CLASS_MASK     (0xFFFF0000)
+#define UVISOR_ERROR_MASK           (0x0000FFFF)
+
+#define UVISOR_ERROR_CLASS_PAGE     (1ul << 16)
```
s/1ul/1UL/

AlessandroA commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,62 @@
+
+/* Must be a power of 2 for MPU alignment in ARMv7-M */
+#ifndef UVISOR_PAGE_SIZE
+#   define UVISOR_PAGE_SIZE ((size_t)16 * 1024)
```
We never indented pre-processor directives. I'd stick to the following for now:

```C
#ifdef
#define
#endif
```

AlessandroA replied:
s/`(size_t)16`/`(size_t) 16`/

AlessandroA commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,62 @@
+
+#include <stdint.h>
+#include <stddef.h>
+#include "halt_exports.h"
```
Please use the full path here (we use this for all API header files). Could you also move the inclusion before the `std` ones? It helps ensuring that `halt_exports.h` is self-contained.

AlessandroA commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,62 @@
+#endif
+
+/* Returns the rounded up number of pages required to hold `size` */
+#define UVISOR_PAGES_FOR_SIZE(size) \
```
This can go in 1 line.

AlessandroA commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,62 @@
+    }
+
+/* Creates a page table with enough pages to hold `size` */
+#define UVISOR_PAGE_TABLE_FOR_SIZE(size) \
```
This can go in 1 line.

AlessandroA commented on core/system/inc/page_allocator.h:
```diff
@@ -0,0 +1,51 @@
+ * This means that not all of the memory between start and end address can be
+ * used by the heap address, due to adhering to alignment restrictions.
+ *
+ * @param heap_start[in] start address of the page heap
```
Capitalize the `@param` comments (3 of them here).

AlessandroA commented on core/system/inc/page_allocator.h:
```diff
@@ -0,0 +1,51 @@
+ */
+int page_allocator_free(const UvisorPageTable * const table);
+
+#endif /*__PAGE_ALLOCATOR_H__*/
```
```diff
-#endif /*__PAGE_ALLOCATOR_H__*/
+#endif /* __PAGE_ALLOCATOR_H__ */
```

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,240 @@
+/* This file can be compiled externally to provide the page allocator algorithm
+   for devices NOT supported by uVisor. For this purpose this file is copied as
+   is into uvisor-mbed-lib and compiled by the target build system. */
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
```
See [here](https://github.com/ARMmbed/uvisor/pull/227/commits/af2a299e9d6f822e1187b50f5333b78175cf848c#r65154640).

niklas-arm commented on core/system/inc/mpu/vmpu.h:
```diff
@@ -162,6 +162,30 @@ extern int vmpu_box_id_self(void);
+ * With the K64F MPU that would be 32 <= 32*N <= 512M.
+ *
+ * @note `size` is limited to 512M, even though larger sizes are allowed by the
+ *       MPU, but don't make sense from a security perspective.
```
because?

AlessandroA commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,60 @@
+#define UVISOR_PAGE_SIZE ((uint32_t) 16 * 1024)
+#endif
+
+/* Returns the rounded up number of pages required to hold `size` */
```
Please use the imperative mood and punctuation:

`/* Return the rounded up number of pages required to hold `size`. */`

(It applies to all occurrences in this PR).

AlessandroA replied:
To clarify, punctuation only applies if the comment is a full English sentence.

```C
/* Maximum context depth */
#define MAX_DEPTH 10
```

```C
/* Check that the value is valid. */
bool is_value_valid(uint32_t value);
```

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,249 @@
+/* By default a maximum of 16 pages are allowed. This can only
+   be decided by the porting engineer for the current platform! */
+#ifndef UVISOR_PAGE_TABLE_MAX_COUNT
+#define UVISOR_PAGE_TABLE_MAX_COUNT ((size_t)16)
```
```diff
-(size_t)16
+(size_t) 16
```

meriac commented on core/system/inc/linker.h:
```diff
@@ -65,6 +65,10 @@ typedef struct {
 
+    /* heap for page allocator */
+    uint32_t * page_start, * page_end;
+    uint32_t page_size;
```
@niklas-arm Isn't that a pointer to page_size ?

niklas-arm replied:
yes… yes, it is…
_cries in corner_

niklas-arm commented:
I added pointer range checks for all input pointers passed to the page allocator init function.
Also added `vmpu_is_region_size_valid` and `vmpu_round_up_region` functions to make the page size and page alignment specialized on the MPU implementation.

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,272 @@
+ */
+
+/* This file can be compiled externally to provide the page allocator algorithm
+   for devices NOT supported by uVisor. For this purpose this file is copied as
```
I would remove the mbed-lib-specific sentence, but the comment is important — if you want to elaborate more in place of the mbed-lib part please do!

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,272 @@
+   is into uvisor-mbed-lib and compiled by the target build system. */
+#if defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1)
+#include <uvisor.h>
+#include "page_allocator.h"
```
How does it work for `page_allocator.h`? It must be copied as well? Other header files (`*_exports.h` ones) are usually available for free (or copied altogether), but this would need to be copied separately.

AlessandroA replied:
Sorry this doesn't apply, my mistake.

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,272 @@
+#endif
+
+/* By default a maximum of 16 pages are allowed. This can only
+   be decided by the porting engineer for the current platform! */
```
```diff
-   be decided by the porting engineer for the current platform! */
+ * be decided by the porting engineer for the current platform! */
```

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,272 @@
+static uint8_t g_page_count_total;
+
+
+void page_allocator_init(void * const heap_start, void * const heap_end, uint32_t * const page_size)
```
I guess page_size is a `uint32_t const * const` (constant pointer to constant uint32)?

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+    const uint32_t mask = size - 1;
+    return (addr + mask) & ~mask;
+}
+static uint32_t page_table_read(uint32_t addr)
```
Can this be inlined?

Patater replied:
It's static. The compiler will likely inline it automatically. Check the disassembly to see.

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+{
+    return *((uint32_t *) addr);
+}
+static void page_table_write(uint32_t addr, uint32_t data)
```
Can this be inlined?

Patater commented on core/system/inc/mpu/vmpu.h:
```diff
@@ -162,6 +162,30 @@ extern int vmpu_box_id_self(void);
 extern int vmpu_box_id_caller(void);
 extern int vmpu_box_namespace_from_id(int box_id, char *box_name, size_t length);
 
+/** Determines if the passed size can be mapped to an exact region size
```
Use imperative mood.

Patater commented on core/system/inc/mpu/vmpu.h:
```diff
@@ -162,6 +162,30 @@ extern int vmpu_box_id_self(void);
+ * @param size Region size chosen for alignment
+ *
+ * @retval >0 The passed address rounded up to the next possible alignment
+ * of the passed region size.
```
Remove period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+
+int vmpu_is_region_size_valid(uint32_t size)
+{
+    /* Get the MSB of the size */
```
Add period

niklas-arm replied:
I don't think adding a period contributes at all to making this code more understandable or maintainable or more secure.
This is an internal comment, not API documentation.

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+        /* 2^29 == 512M, which is the maximum region size */
+        return 0;
+    }
+    /* Compute the size from MSB */
```
Add period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+    }
+    /* Compute the size from MSB */
+    const int bit_size = (1UL << bits);
+    /* There is no round up */
```
Add period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+    const uint32_t mask = size - 1;
+    /* Mask is 31 <= mask <= (1 << 29) - 1 */
+
+    /* Adding the mask can overflow */
```
Add period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+
+    /* Adding the mask can overflow */
+    const uint32_t rounded_addr = addr + mask;
+    /* Check for overflow */
```
Add period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+    const uint32_t rounded_addr = addr + mask;
+    /* Check for overflow */
+    if (rounded_addr < addr) {
+        /* This means the address was too large to align */
```
Add period

Patater commented on core/system/src/mpu/vmpu_armv7m.c:
```diff
@@ -717,3 +717,42 @@ void vmpu_arch_init_hw(void)
+        /* This means the address was too large to align */
+        return 0;
+    }
+    /* Mask the rounded address to get the aligned address */
```
Add period

Patater commented on core/system/src/mpu/vmpu_freescale_k64.c:
```diff
@@ -296,3 +295,36 @@ void vmpu_arch_init(void)
+        /* 2^29 == 512M, which is the maximum region size */
+        return 0;
+    }
+    /* There is no rounding, we only care about an exact match! */
```
s/`,`/`.`
s/w/W

niklas-arm replied:
`s/,/;`?

Patater commented on core/system/src/mpu/vmpu_freescale_k64.c:
```diff
@@ -296,3 +295,36 @@ void vmpu_arch_init(void)
+
+int vmpu_is_region_size_valid(uint32_t size)
+{
+    /* Align size to 32B */
```
Add period

Patater commented on core/system/src/mpu/vmpu_freescale_k64.c:
```diff
@@ -296,3 +295,36 @@ void vmpu_arch_init(void)
+    /* Alignment is always 32B */
+    const uint32_t mask = 31;
+
+    /* Adding the mask can overflow */
```
Add period

niklas-arm replied:
YES, I GET IT.

niklas-arm replied:
The period is clearly relevant to the security of this code.

Patater commented on core/system/src/mpu/vmpu_freescale_k64.c:
```diff
@@ -296,3 +295,36 @@ void vmpu_arch_init(void)
+    return (masked_size == size);
+}
+
+uint32_t vmpu_round_up_region(uint32_t addr, uint32_t size)
```
Can the implementation of `vmpu_round_up_region` be shared between these two MPUs? The code looks pretty much the same, other than which `vmpu_region_size_valid` function to call.

niklas-arm replied:
I thought so too, but the mask is different. On ARMv7-M the mask is the (region size - 1), on K64F it's always (32 - 1), regardless of region size.

Patater commented:
Some commit organization things
 - Remove period from end of `Add is_region_valid and round_up_region to vmpu.`.
 - Squash `Add page allocator exports` with `Add page allocator SVC calls for API` and `Add page allocator implementation`. It's a three-way squash. All commits provide implementation of the page allocator API.
- Squash `Add page heap start and end address to config` into `Initialize page heap with start and end address`. It's implementation detail for initializing the page heap.

Patater commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,60 @@
+
+
+#define UVISOR_ERROR_PAGE_OK                    (0)
+#define UVISOR_ERROR_PAGE_OUT_OF_MEMORY         (UVISOR_ERROR_CLASS_PAGE + 100)
```
Since you have error classes now, it probably isn't necessary to start at 100, but if you want to that is fine.

Patater commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,60 @@
+#define UVISOR_ERROR_PAGE_INVALID_PAGE_COUNT    (UVISOR_ERROR_CLASS_PAGE + 105)
+
+
+/* Must be a power of 2 for MPU alignment in ARMv7-M. */
```
Add a static assert to check this alignment.

niklas-arm replied:
This is actually half true.
It must be power of 2 for the ARM MPU, but can be 32B-multiple for K64F MPU.

niklas-arm replied:
I can't add a static check. The page size is passed through the uvisor-input.S, so can only be checked at runtime.

Patater replied:
It would be a static check for programmer sanity, not security. We'd still need the check inside uVisor. Having a static check here earlier would help the programmer catch his mistake sooner and in an obvious way: he wouldn't have to pull out the debugger to see the uVisor halt sanity check failed message.

niklas-arm replied:
Yes, but the static check would need to be specialized on MPU type.
ie. power of two for ARM MPU, 32 multiple for K64F.
Not sure if we want that in a header file.

Patater replied:
Nah, wouldn't want that specialization just for a static assert. We could consider forcing everybody to use the least common denominator MPU capabilities (forcing everybody to use power of 2 alignment so their code is portable), but that's probably not worth it.

At least update your comment to reflect reality, or delete it so it doesn't confuse anybody. We'll have to provide more detailed rules about setting the `UVISOR_PAGE_SIZE` in our porting guide.

Patater commented on api/inc/page_allocator_exports.h:
```diff
@@ -0,0 +1,60 @@
+typedef struct {
+    uint32_t page_size;     /* The page size in bytes */
+    uint32_t page_count;    /* The number of pages in the page table */
+    void * page_origins[1]; /* Table of pointer to the origins of a page */
```
s/pointer/pointers

Because of `origins`, this reads as if there are multiple origins per page. Is that correct?

Patater commented on core/system/inc/page_allocator.h:
```diff
@@ -0,0 +1,51 @@
+void page_allocator_init(void * const heap_start, void * const heap_end, const uint32_t * const page_size);
+
+/* Allocate a number of requested pages with the requested page size.
+ * @param table.page_size[in]     Must be equal to the current page size.
```
Remove period.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+#ifndef UVISOR_PAGE_TABLE_MAX_COUNT
+#define UVISOR_PAGE_TABLE_MAX_COUNT ((size_t) 16)
+#endif
+/* The page box_id is the box id which is 8bit large  */
```
s/8bit/8-bit

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+#endif
+/* The page box_id is the box id which is 8bit large  */
+typedef uint8_t page_owner_t;
+/* Maps the page to the owning box handle */
```
Use imperative mood.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+typedef uint8_t page_owner_t;
+/* Maps the page to the owning box handle */
+static page_owner_t g_page_owner_table[UVISOR_PAGE_TABLE_MAX_COUNT];
+/* define a unused value for the page table */
```
Capitalize comment beginnings.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+            "Page heap end pointer (0x%08x) is not in sram memory.\n",
+            (unsigned int) heap_end);
+    }
+    if (heap_end < heap_start) {
```
Is having `heap_end == heap_start` (a size zero heap) not allowed?

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+            (unsigned int) heap_end,
+            (unsigned int) heap_start);
+    }
+    if ((*page_size < 1024) || !vmpu_is_region_size_valid(*page_size)) {
```
What went into the decision to make 1024 the minimum page size? Add a comment in the code.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+        return 0;
+    }
+    const uint32_t mask = size - 1;
+    return (addr + mask) & ~mask;
```
What would happen if adding mask to addr overflowed?

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    g_page_size = *page_size;
+    /* This is the page heap start address */
```
Add period

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+    if ((start + g_page_size) > (uint32_t) heap_end) {
+        g_page_count_total = 0;
+    }
+    else {
```
Put the `}` on the same line as the `else {`.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+    else {
+        g_page_count_total = ((uint32_t) heap_end - start) / g_page_size;
+    }
+    /* Clamp page count to table size */
```
Grammars... :(

I'll stop bothering you about it.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+#endif /* defined(UVISOR_PRESENT) && (UVISOR_PRESENT == 1) */
+
+/* By default a maximum of 16 pages are allowed. This can only
+ * be decided by the porting engineer for the current platform! */
```
Add comments about the rationale or things a porting engineer must consider when decided on this value. Explain the rationale behind choosing 16.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+        /* If the page is unused, it's entry is UVISOR_PAGE_UNUSED (not NULL!) */
+        if (g_page_owner_table[page] == UVISOR_PAGE_UNUSED) {
+            /* Marry this page to the box id */
+            g_page_owner_table[page] = box_id;
```
Is it possible that another call to `page_allocator_malloc` in a different thread or interrupt could claim a page after our check, but before we set it for this box_id? If this function assumes a lock is held before being called, please add a comment explaining the assumptions. If we don't assume a lock is held, we could use an atomic test and set built-in if any re-entrancy is required.

niklas-arm replied:
This assumes being called in an SVC, which makes this mutually exclusive.
But, since this same file is used for unsupported, this does indeed lead to a race condition.

niklas-arm replied:
Going to implement a mutex for alloc and free for unsupported only.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+        size_t page_index = (page - g_page_heap_start) / g_page_size;
+        /* Check if the page belongs to the caller */
+        if (g_page_owner_table[page_index] == box_id) {
+            free_mask[page_index/32] |= (1 << (page_index & 31));
```
Place spaces around arithmetic operators, like `/`.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,287 @@
+        }
+    }
+    /* We have freed some pages */
+    g_page_count_free += page_count;
```
Updates to `g_page_count_free` must also be atomic.

niklas-arm commented:
Addressed review and squashed commits.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,333 @@
+    /* Iterate over the bits in the free mask and actually free the pages. */
+    uint32_t count = 0;
+    for (; count < g_page_count_total; count++) {
+        if (free_mask[count/32] & (1 << (count & 31))) {
```
Place spaces around arithmetic operators, like `/`.

Patater commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,333 @@
+/* Mutex implementation must be provided by the OS, since the page allocator
+ * is called directly from user code instead of through an SVC call. */
+#ifndef UVISOR_PAGE_ALLOCATOR_MUTEX_AQUIRE
+#define UVISOR_PAGE_ALLOCATOR_MUTEX_AQUIRE {}
```
Should we `#error` here instead of silently failing to acquire a mutex if the uVisor porting engineer forgot to implement this macro?

niklas-arm replied:
Moved to https://github.com/ARMmbed/uvisor/pull/232/commits/2d9ae6bc52a54f919c65881e01d604135aa61f68#diff-c71a307c4fd911388ca9a0399570ac88R99

meriac commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+    /* Update the free pages count. */
+    g_page_count_free -= pages_required;
+    /* Point to the first entry in the table. */
+    void * * page_table = table->page_origins;
```
@niklas-arm page_table_read ?

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+    g_page_heap_start = (void *) start;
+
+    /* How many pages can we fit in here? */
+    if ((start + g_page_size) > (uint32_t) heap_end) {
```
Is this guaranteed not to overflow?

niklas-arm replied:
Good point, fixed.

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+    return UVISOR_ERROR_PAGE_OK;
+}
+
+int page_allocator_free(const UvisorPageTable *const table)
```
`* const`

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+    void * const * page_table = table->page_origins;
+
+    for (; table_size > 0; page_table++, table_size--) {
+        void *page = (void *) page_table_read((uint32_t) page_table);
```
`void * page`

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+        /* Check if the page belongs to the caller. */
+        if (g_page_owner_table[page_index] == box_id) {
+            free_mask[page_index / 32] |= (1 << (page_index & 31));
+        }
```
```diff
-        }
-        /* Abort if the page doesn't belong to the caller. */
-        else {
-            if (g_page_owner_table[page_index] == UVISOR_PAGE_UNUSED) {

+        } else {
+            /* Abort if the page doesn't belong to the caller. */
+            if (g_page_owner_table[page_index] == UVISOR_PAGE_UNUSED) {
```

AlessandroA commented on core/system/src/page_allocator.c:
```diff
@@ -0,0 +1,285 @@
+    /* Iterate over the table and validate each pointer. */
+    void * const * page_table = table->page_origins;
+
+    for (; table_size > 0; page_table++, table_size--) {
```
Can you move `int table_size = page_count;` here (right before the loop)?

AlessandroA commented on api/src/uvisor-input.S:
```diff
@@ -84,6 +85,11 @@ uvisor_config:
     /* pointer to __uvisor_box_context */
     .long __uvisor_box_context
 
+    /* start and end address and page size of page heap */
```
Can you replace this comment with something like `/* Page allocator heap boundaries and size */`? See [this commit](https://github.com/ARMmbed/uvisor/pull/233/commits/d1128a81846fed3f584b8677bbd46393a8ae651d) for reference.

AlessandroA commented on core/system/inc/linker.h:
```diff
@@ -65,6 +65,10 @@ typedef struct {
     /* address of __uvisor_box_context */
     uint32_t **uvisor_box_context;
 
+    /* heap for page allocator */
```
Can you replace this comment with something like `/* Page allocator heap boundaries and size */`? See [this commit](https://github.com/ARMmbed/uvisor/pull/233/commits/d1128a81846fed3f584b8677bbd46393a8ae651d) for reference.

AlessandroA commented on core/system/inc/linker.h:
```diff
@@ -65,6 +65,10 @@ typedef struct {
     uint32_t **uvisor_box_context;
 
+    /* heap for page allocator */
+    uint32_t * page_start, * page_end;
```
Please move on 2 lines for consistency with [this commit](https://github.com/ARMmbed/uvisor/pull/233/commits/d1128a81846fed3f584b8677bbd46393a8ae651d):

```
uint32_t * page_start;
uint32_t * page_end;
```
Thanks!
